File: Napakalaki Ruby/CardDealer.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.


require_relative 'Monster.rb'
require_relative 'Treasure.rb'
require_relative 'NumericBadConsequence.rb'
require_relative 'DeathBadConsequence.rb'
require_relative 'SpecificBadConsequence.rb'
require_relative 'Cultist.rb'

module NapakalakiGame

class CardDealer
  
  attr_reader :unused_monsters, :used_monsters, :used_treasures, :unused_treasures, :unusedCultists
  
  @@instance = nil
  private_class_method :new
  
 private
 
  def initialize
    @unused_monsters = Array.new
    @used_monsters = Array.new
    @unused_treasures = Array.new
    @used_treasures = Array.new
    @unused_cultist = Array.new
  end
  
  def shuffleCultists
    
    @unused_cultist.shuffle!
  end
  
  def initCultistCardDeck
    @unused_cultist = Array.new
    
    @unused_cultist << Cultist.new('Sectario',1)
    @unused_cultist << Cultist.new('Sectario',2)
    @unused_cultist << Cultist.new('Sectario',1)
    @unused_cultist << Cultist.new('Sectario',2)
    @unused_cultist << Cultist.new('Sectario',1)
    @unused_cultist << Cultist.new('Sectario',1)
   
    shuffleCultists
    
  end
  
  def initTreasureCardDeck ()
    treasure_deck = Array.new
    
    treasure_deck << Treasure.new("Si mi amo",4,TreasureKind::HELMET)
    treasure_deck << Treasure.new("Botas de investigacion",3,TreasureKind::SHOES)
    treasure_deck << Treasure.new("Capucha de Cthulhu",3,TreasureKind::HELMET)
    treasure_deck << Treasure.new("A prueba de babas",2,TreasureKind::ARMOR)
    treasure_deck << Treasure.new("Botas de lluvia acida",1,TreasureKind::BOTHHANDS)
    treasure_deck << Treasure.new("Casco minero",2,TreasureKind::HELMET)
    treasure_deck << Treasure.new("Ametralladora Thompson",4,TreasureKind::BOTHHANDS)
    treasure_deck << Treasure.new("Camiseta de la UGR",1,TreasureKind::ARMOR)
    treasure_deck << Treasure.new("Clavo de rail ferroviario",3,TreasureKind::ONEHAND)
    treasure_deck << Treasure.new("Cuchillo de sushi arcano",2,TreasureKind::ONEHAND)
    treasure_deck << Treasure.new("Hacha prehistorica",2,TreasureKind::ONEHAND)
    treasure_deck << Treasure.new("El aparato del Pr.Tesla",4,TreasureKind::ARMOR)
    treasure_deck << Treasure.new("Gaita",4,TreasureKind::BOTHHANDS)
    treasure_deck << Treasure.new("Insecticida",2,TreasureKind::ONEHAND)
    treasure_deck << Treasure.new("Escopeta de tres caniones",4,TreasureKind::BOTHHANDS)
    treasure_deck << Treasure.new("Garabato mistico",2,TreasureKind::ONEHAND)
    treasure_deck << Treasure.new("La rebeca metalica",2,TreasureKind::ARMOR)
    treasure_deck << Treasure.new("Lanzallamas",4,TreasureKind::BOTHHANDS)
    treasure_deck << Treasure.new("Necrocomicon",1,TreasureKind::ONEHAND)
    treasure_deck << Treasure.new("Necronomicon",5,TreasureKind::BOTHHANDS)
    treasure_deck << Treasure.new("Linterna a 2 manos",3,TreasureKind::BOTHHANDS)
    treasure_deck << Treasure.new("Necrognomicon",2,TreasureKind::ONEHAND)
    treasure_deck << Treasure.new("Necrotelecom",2,TreasureKind::HELMET)
    treasure_deck << Treasure.new("Mazo de los atiguos",3,TreasureKind::ONEHAND)
    treasure_deck << Treasure.new("Necroplayboycon",3,TreasureKind::ONEHAND)
    treasure_deck << Treasure.new("Porra preternatural",2,TreasureKind::ONEHAND)
    treasure_deck << Treasure.new("Shogulador",1,TreasureKind::BOTHHANDS)
    treasure_deck << Treasure.new("Varita de atizamiento",3,TreasureKind::ONEHAND)
    treasure_deck << Treasure.new("Tentaculo de pega",2,TreasureKind::HELMET)
    treasure_deck << Treasure.new("Zapato deja-amigos",1,TreasureKind::SHOES)
    @unused_treasures = treasure_deck
    shuffleTreasures
  end
  
  def initMonsterCardDeck ()
    monsters = Array.new

    prize = Prize.new(2,1)
    badConsequence = 
      SpecificBadConsequence.new('Pierdes tu armadura visible y otra oculta',0,[TreasureKind::ARMOR],[TreasureKind::ARMOR])
    monsters << Monster.new('3 Byakhees de bonanza',8,badConsequence,prize,0)


    prize = Prize.new(4,2)
    badConsequence =
    NumericBadConsequence.new('Pierdes 5 niveles y 3 tesoros visibles',5 , 3, 0)
    monsters << Monster.new('El rey de rosa',13,badConsequence,prize,0)


    prize = Prize.new(1,1)
    badConsequence = 
      SpecificBadConsequence.new('Embobados con el lindo primigenio te descartas de tu casco visible',0,[TreasureKind::HELMET],[])
    monsters << Monster.new('Chibithulhu',2,badConsequence,prize,0)


    prize = Prize.new(1,1)
    badConsequence = 
      SpecificBadConsequence.new('El primordial bostezo contagioso.Pierdes el calzado visible',0,[TreasureKind::SHOES],[])
    monsters << Monster.new('El sopor de Dunwich',2,badConsequence,prize,0)
  
 
    prize = Prize.new(4,1)
    badConsequence = 
      SpecificBadConsequence.new('Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y 1 mano oculta',0,[TreasureKind::ONEHAND],[TreasureKind::ONEHAND])
    monsters<< Monster.new('Angeles de la noche ibicenca', 14, badConsequence,prize,0)


    prize = Prize.new(3,1)
    badConsequence = 
        NumericBadConsequence.new('Pierdes todos tus tesoros visibles',0,5,0)
    monsters<< Monster.new('El gorron en el umbral',10,badConsequence,prize,0)


    prize = Prize.new(2,1)
    badConsequence = 
        SpecificBadConsequence.new('Pierdes la armadura visible',0,[TreasureKind::ARMOR],[])
    monsters<< Monster.new('H.P.Munchcraft',6,badConsequence,prize,0)


    prize = Prize.new(1,1)
    badConsequence = 
        SpecificBadConsequence.new('Sientes bichos bajo la ropa.Descarta la armadura visible',0,[TreasureKind::ARMOR],[])
    monsters<< Monster.new('Bichgooth',2,badConsequence,prize,0)


    prize = Prize.new(1,1)
    badConsequence = 
      NumericBadConsequence.new('Toses los pulmones y pierdes 2 niveles',2,0,0)
    monsters << Monster.new('La que redacta en las tinieblas',2,badConsequence,prize,0)


    prize = Prize.new(2,1)
    badConsequence = 
      DeathBadConsequence.new('Estos monstruos resultan bastante superficiales y te aburren mortalmente. estas muerto',true)
    monsters << Monster.new('Los hondos',8,badConsequence,prize,0)


    prize = Prize.new(2,1)
    badConsequence = 
      NumericBadConsequence.new('Pierdes 2 niveles y 2 tesoros ocultos',2,0,2)
    monsters << Monster.new('Semillas Cthulhu',4,badConsequence,prize,0)


    prize = Prize.new(2,1)
    badConsequence = 
        SpecificBadConsequence.new('Te intentas escaquear. Pierdes una mano visible',0,[TreasureKind::ONEHAND],[])
    monsters<< Monster.new('Dameargo',1,badConsequence,prize,0)


    prize = Prize.new(1,1)
    badConsequence = 
      NumericBadConsequence.new('Da mucho asquito. Pierdes 3 niveles',3,0,0)
    monsters << Monster.new('Pollipolipo volante',3,badConsequence,prize,0)


    prize = Prize.new(3,1)
    badConsequence = 
      DeathBadConsequence.new('No le hace gracia que pronuncien mal su nombre. Estas muerto',true)
    monsters << Monster.new('Yskhtihyssg-Goth',12,badConsequence,prize,0)


    prize = Prize.new(4,1)
    badConsequence = 
      DeathBadConsequence.new('La familia te atrapa. Estas muerto',true)
    monsters << Monster.new('Familia feliz',1,badConsequence,prize,0)


    prize = Prize.new(2,1)
    badConsequence = 
        SpecificBadConsequence.new('La quinta directiva primaria te obliga a perder 2 niveles y un tesoro 2 manos visible',2,[TreasureKind::BOTHHANDS],[])
    monsters<< Monster.new('Roboggoth',8,badConsequence,prize,0)


    prize = Prize.new(1,1)
    badConsequence = 
        SpecificBadConsequence.new('Te asusta en la noche. Pierdes un casco visible',0,[TreasureKind::HELMET],[])
    monsters<< Monster.new('El espia',5,badConsequence,prize,0)


    prize = Prize.new(1,1)
    badConsequence = 
      NumericBadConsequence.new('Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles',2,5,0)
    monsters << Monster.new('El Lenguas',20,badConsequence,prize,0)


    prize = Prize.new(1,1)
    badConsequence = 
      SpecificBadConsequence.new('Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos',3,[TreasureKind::ONEHAND,TreasureKind::BOTHHANDS],[])
    monsters<< Monster.new('Bicefalo',20,badConsequence,prize,0)
    
    prize = Prize.new(3,1)
    badConsequence = 
      SpecificBadConsequence.new('Pierdes 1 mano visible',0,[TreasureKind::ONEHAND],[])
    monsters<< Monster.new('El mal indecible impronunciable',10,badConsequence,prize,-2)
    
    prize = Prize.new(2,1)
    badConsequence = 
      NumericBadConsequence.new('Pierdes tus tesoros visibles',0,5,0)
    monsters << Monster.new('Testigos oculares',6,badConsequence,prize,2)
    
    prize = Prize.new(2,5)
    badConsequence = 
      DeathBadConsequence.new('Hoy no es tu dia de suerte.Mueres',true)
    monsters << Monster.new('El gran cthulhu',20,badConsequence,prize,4)
    
    prize = Prize.new(2,1)
    badConsequence = 
      NumericBadConsequence.new('Tu gobierno te recorta 2 niveles',2,0,0)
    monsters << Monster.new('Serpiente politico',8,badConsequence,prize,-2)
    
    prize = Prize.new(1,1)
    badConsequence = 
      SpecificBadConsequence.new('Pierdes tu caso y tu armadura visible.Pierdes tus manos ocultas',0,[TreasureKind::HELMET,TreasureKind::ARMOR],[TreasureKind::BOTHHANDS])
    monsters<< Monster.new('Felpuggoth',2,badConsequence,prize,5)
    
    prize = Prize.new(4,2)
    badConsequence = 
      NumericBadConsequence.new('Pierdes 2 niveles',2,0,0)
    monsters << Monster.new('Shoggoth',16,badConsequence,prize,-4)
    
    prize = Prize.new(1,1)
    badConsequence = 
      NumericBadConsequence.new('Pintalabios negro.Pierdes 2 niveles',2,0,0)
    monsters << Monster.new('Lolitagooth',2,badConsequence,prize,3)
    
    @unused_monsters = monsters
    shuffleMonsters
  end
  
  def shuffleTreasures()
    @unused_treasures.shuffle!
  end
  
  def shuffleMonsters ()
    @unused_monsters.shuffle!
  end
  
  public
  
  def nextCultist
    if(@unused_cultist.empty?)
      aux = nil
    else
      aux = @unused_cultist.shift
    end 
    return aux
  end
  
  def self.instance ()
    @@instance = new if (@@instance == nil)
    return @@instance
  end
  
  def nextTreasure ()
    if ( @unused_treasures.empty?)
      @unused_treasures = @used_treasures
      @used_treasures = Array.new
      @unused_treasures.shuffle!
    end
      
      sig = @unused_treasures.shift
    
    return sig
  end
  
  def nextMonster ()
    if (@unused_monsters.empty?)
      @unused_monsters = @used_monsters
      @used_monsters = Array.new
      @unused_monsters.shuffle!
    end
      
      sig = @unused_monsters.shift
    
    return sig    
    
  end
  
  def giveTreasureBack (t) #Treasure t
    @used_treasures << t
  end
  
  def giveMonsterBack (m) #Monster m
    @used_monsters << m
  end
  
  def initCards ()
    initTreasureCardDeck
    initMonsterCardDeck
    initCultistCardDeck
  end
end

end

File: Napakalaki Java/NapakalakiGame/CardDealer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import static NapakalakiGame.TreasureKind.ARMOR;
import static NapakalakiGame.TreasureKind.BOTHHANDS;
import static NapakalakiGame.TreasureKind.HELMET;
import static NapakalakiGame.TreasureKind.ONEHAND;
import static NapakalakiGame.TreasureKind.SHOES;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
/**
 *
 * @author Antonio
 */
public class CardDealer {
    
    ////////////////////////////////////////
    /////////////////////////////////////////
    
    private static CardDealer dealer = null;
    private ArrayList<Monster> usedMonsters = null;
    private ArrayList<Monster> unusedMonsters = null;
    private ArrayList<Treasure> usedTreasures = null;
    private ArrayList<Treasure> unusedTreasures = null;
    private ArrayList<Cultist> unusedCultists = null;
    
    
    private CardDealer(){
        //Para crear un objeto en el main:
        //CardDealer dealer = CardDealer.getInstance();
        usedMonsters = new ArrayList();
        unusedMonsters = new ArrayList();
        usedTreasures = new ArrayList();
        unusedTreasures = new ArrayList();
        unusedCultists = new ArrayList();
        
    }
    
    private void initTreasureCardDeck(){
        
        unusedTreasures.add(new Treasure("¡Si mi amo!", 4, HELMET));
        unusedTreasures.add(new Treasure("Botas de investigacion", 3, SHOES));
        unusedTreasures.add(new Treasure("Capucha de Cthulhu", 3, HELMET));
        unusedTreasures.add(new Treasure("A prueba de babas", 2, ARMOR));
        unusedTreasures.add(new Treasure("Botas de lluvia acida", 1, BOTHHANDS));
        unusedTreasures.add(new Treasure("Casco minero", 2, HELMET));
        unusedTreasures.add(new Treasure("Ametralladora Thompson", 4, BOTHHANDS));
        unusedTreasures.add(new Treasure("Camiseta de la UGR", 1, ARMOR));
        unusedTreasures.add(new Treasure("Clavo de rail ferroviario", 3, ONEHAND));
        unusedTreasures.add(new Treasure("Cuchillo de sushi arcano", 2, ONEHAND));
        unusedTreasures.add(new Treasure("Fez alopodo", 3, HELMET));
        unusedTreasures.add(new Treasure("Hacha prehistorica", 2, ONEHAND));
        unusedTreasures.add(new Treasure("El aparato del PR.Tesla", 4, ARMOR));
        unusedTreasures.add(new Treasure("Gaita", 4, BOTHHANDS));
        unusedTreasures.add(new Treasure("Insecticida", 2, ONEHAND));
        unusedTreasures.add(new Treasure("Escopeta de 3 canones", 4, BOTHHANDS));
        unusedTreasures.add(new Treasure("Garabato mistico", 2, ONEHAND));
        unusedTreasures.add(new Treasure("La rebeca metalica", 2, ARMOR));
        unusedTreasures.add(new Treasure("Lanzallamas", 4, BOTHHANDS));
        unusedTreasures.add(new Treasure("Necro-comicon", 1, ONEHAND));
        unusedTreasures.add(new Treasure("Necronomicon", 5, BOTHHANDS));
        unusedTreasures.add(new Treasure("Linterna a 2 manos", 3, BOTHHANDS));
        unusedTreasures.add(new Treasure("Necro-gnomicon", 2, ONEHAND));
        unusedTreasures.add(new Treasure("Necrotelecom", 2, HELMET));
        unusedTreasures.add(new Treasure("Mazo de los antiguos", 3, ONEHAND));
        unusedTreasures.add(new Treasure("Necro-playboycon", 3, ONEHAND));
        unusedTreasures.add(new Treasure("Porra preternatural", 2, ONEHAND));
        unusedTreasures.add(new Treasure("Shogulador", 1, BOTHHANDS));
        unusedTreasures.add(new Treasure("Varita de atizamiento", 3, ONEHAND));
        unusedTreasures.add(new Treasure("Tentaculo de pega", 2, HELMET));
        unusedTreasures.add(new Treasure("Zapato deja-amigos", 1, SHOES));
        
         shuffleTreasures();
        
    }
    
    private void initMonsterCardDeck(){
        
        BadConsequence mal_rollo;
        
        //3 Byakhees de bonanza 
        mal_rollo = new SpecificBadConsequence("Pierdes tu armadura visible y otra oculta", 0, 
                    new ArrayList(Arrays.asList(TreasureKind.ARMOR)),
                    new ArrayList(Arrays.asList(TreasureKind.ARMOR)));
        
        unusedMonsters.add(new Monster("3 Byakhees de bonanza", 8, mal_rollo, new Prize(2,1)));
        
        //Chibithulhu 
        mal_rollo = new SpecificBadConsequence("Embolados con el lindo primigenio te descartas de tu casco visible",
                        0,new ArrayList(Arrays.asList(TreasureKind.HELMET)),new ArrayList());
        unusedMonsters.add(new Monster("Chibithulhu", 2, mal_rollo, new Prize(1,1)));
        
        //El sopor de Dunwich 
        mal_rollo = new SpecificBadConsequence("El primordial bostezo contagioso. Pierdes el calzado visible", 0,
                    new ArrayList (Arrays.asList(TreasureKind.SHOES)),
                    new ArrayList());
        unusedMonsters.add(new Monster("El sopor de Dunwich", 2, mal_rollo, new Prize(1,1)));
        
        //Angeles de la noche ibicenca
        mal_rollo = new SpecificBadConsequence("Te atrapan para llevarte de fiesta y te dejan caer en mitad de vuelo. Descarta una mano visible y una mano coulta",
                                        0,new ArrayList(Arrays.asList(TreasureKind.ONEHAND)),new ArrayList(Arrays.asList(TreasureKind.ONEHAND)));
        
        unusedMonsters.add(new Monster("Angeles de la noche ibicenca",14, mal_rollo, new Prize(4,1)));
        
        //El gorron en el humbral
        mal_rollo = new NumericBadConsequence("Pierdes todos tus tesoros visibles",0,5,0);
        
        unusedMonsters.add(new Monster("El gorron en el humbral", 10, mal_rollo, new Prize(3,1)));
        
        //H.P. Munchcraft
        mal_rollo = new SpecificBadConsequence("Pierdes la armadura visible",0,
                    new ArrayList (Arrays.asList(TreasureKind.ARMOR)),new ArrayList());
        
        unusedMonsters.add(new Monster("H.P. Munchcraft", 6, mal_rollo, new Prize(2,1)));
        
        //Bichgooth
        mal_rollo = new SpecificBadConsequence("Sientes bichos bajo la ropa, descarta la armadura visible", 0,
                    new ArrayList (Arrays.asList(TreasureKind.ARMOR)),new ArrayList());
        
        unusedMonsters.add(new Monster("Bichgooth", 2, mal_rollo, new Prize(1,1)));
        
        //El rey de la rosa
        mal_rollo = new NumericBadConsequence("Pierdes 5 nieveles y tres tesoros visibles",5,3,0);
        
        unusedMonsters.add(new Monster("El rey de la rosa", 13, mal_rollo, new Prize(4,2)));
        
        //La que redacta en las tinieblas
        mal_rollo = new NumericBadConsequence("Toses los pulmones y pierdes dos niveles",2,0,0);
        
        unusedMonsters.add(new Monster("La que redacta en las tinieblas", 2, mal_rollo, new Prize(1,1)));
        
        //Los Hondos
        mal_rollo = new DeathBadConsequence("Estos mosntruos son bastante superficiales y te aburren mortalmente. Estas muerto",0);
        
        unusedMonsters.add(new Monster ("Los hondos", 8, mal_rollo, new Prize(2,1)));
        
        //Semillas Cthulhu
        mal_rollo = new NumericBadConsequence("Pierdes dos niveles y dos tesoros ocultos",2,0,2);
        
        unusedMonsters.add(new Monster("Semillas Cthulhu", 4, mal_rollo, new Prize(2,1)));
        
        //Dameargo
        mal_rollo = new SpecificBadConsequence("Te intentas escaquear. Pierdes una mano visible",0,
                    new ArrayList(Arrays.asList(TreasureKind.ONEHAND)),new ArrayList());
        
        unusedMonsters.add(new Monster("Dameargo",1, mal_rollo, new Prize(2,1)));
        
        //Pollipolipo volante
        mal_rollo = new NumericBadConsequence("Da mucho asquito. Pierdes tres niveles",3,0,0);
        
        unusedMonsters.add(new Monster("Pollipolipo volante", 3, mal_rollo, new Prize(1,1)));
        
        //Yskhtihyssg-Goth
        mal_rollo = new DeathBadConsequence("No le hace gracia que pronuncies mal su nombre. Estas muerto",0);
        
        unusedMonsters.add(new Monster("Yskhtihyssg-Goth", 12, mal_rollo, new Prize(3,1)));
        
        //Familia Feliz
        mal_rollo = new DeathBadConsequence("La familia te atrapa. Estas muerto",0);
        
        unusedMonsters.add(new Monster("La familia feliz", 1, mal_rollo, new Prize(4,1)));
        
        //Roboggoth
        mal_rollo = new SpecificBadConsequence("La quinta directiva primrai te obliga a perder dos niveles y un tesoro de dos manos visible",
                    2, new ArrayList (Arrays.asList(TreasureKind.BOTHHANDS)),new ArrayList());
        
        unusedMonsters.add(new Monster("Roboggoth", 8, mal_rollo, new Prize(2,1)));
        
        //El espia
        mal_rollo = new SpecificBadConsequence("Te asusta en la noche, pierdes un casco visible. Pierdes un casco visible",0,
                    new ArrayList (Arrays.asList(TreasureKind.HELMET)),new ArrayList());
        
        unusedMonsters.add(new Monster("El espia", 5, mal_rollo, new Prize(1,1)));
        
        //El lenguas
        mal_rollo = new NumericBadConsequence("Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles",2,5,0);
         
        unusedMonsters.add(new Monster("El lenguas", 20, mal_rollo, new Prize(1,1)));
        
        //Bicefalo
        mal_rollo = new SpecificBadConsequence("Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las manos",3,
                    new ArrayList (Arrays.asList(TreasureKind.ONEHAND,TreasureKind.BOTHHANDS)),new ArrayList());
         
        unusedMonsters.add(new Monster("Bicefalo", 20, mal_rollo, new Prize(1,1)));
        
        //El mal indecible impronunciable
        mal_rollo = new SpecificBadConsequence("Pierdes 1 mano visible",0,
                                        new ArrayList (Arrays.asList(TreasureKind.ONEHAND)),new ArrayList());
        unusedMonsters.add(new Monster("El mal indecible impronunciable",10,mal_rollo,new Prize(3,1),-2));
        
        //Testigos oculares
        mal_rollo = new NumericBadConsequence("Pierdes tus tesoros visibles. Jajaja",0,5,0);
        unusedMonsters.add(new Monster("Testigos oculares",6,mal_rollo,new Prize(2,1),2));
        
        //El gran cthulhu
        mal_rollo = new DeathBadConsequence("Hoy no es tu dia de suerte, mueres",0);
        unusedMonsters.add(new Monster("El gran cthulhu",20,mal_rollo, new Prize(2,5),4));
        
        //Serpiente Politico
        mal_rollo = new NumericBadConsequence("Tu gobierno te recorta 2 niveles",2,0,0);
        unusedMonsters.add(new Monster("Serpiente Politico",8,mal_rollo,new Prize(2,1),-2));
        
        //Felpuggoth
        mal_rollo = new SpecificBadConsequence("Pierdes tu casco y tu armadura visible. Pierdes tus manos ocultas",0,
                                        new ArrayList (Arrays.asList(TreasureKind.HELMET,TreasureKind.ARMOR)),
                                        new ArrayList (Arrays.asList(TreasureKind.ONEHAND,TreasureKind.ONEHAND,TreasureKind.BOTHHANDS)));
        unusedMonsters.add(new Monster("Felpuggoth",2,mal_rollo,new Prize(1,1),4));
        
        //Shoggoth
        mal_rollo = new NumericBadConsequence("Pierdes dos niveles",2,0,0);
        unusedMonsters.add(new Monster("Shoggoth",16,mal_rollo,new Prize(4,2),-4));
        
        //Lolitagooth
        mal_rollo = new NumericBadConsequence("Pintalabios negro. Pierdes dos niveles",2,0,0);
        unusedMonsters.add(new Monster("Lolitagooth",2,mal_rollo,new Prize(1,1),3));
        
        shuffleMonsters();
    }
    
    private void initCultistCardDeck() {
        Cultist sectario;
        
        sectario = new Cultist("Sectario",1);
        unusedCultists.add(sectario);
        unusedCultists.add(sectario);
        unusedCultists.add(sectario);
        unusedCultists.add(sectario);
        sectario = new Cultist("Sectario",2);
        unusedCultists.add(sectario);
        unusedCultists.add(sectario);
        shuffleCultists();
    }
    
    private void shuffleCultists() {
        Collections.shuffle(unusedCultists);
    }
    
    private void shuffleTreasures(){
        
        Collections.shuffle(unusedTreasures);
    }
    
    private void shuffleMonsters(){
        
        Collections.shuffle(unusedMonsters);
    }
    
    /////////////////////////////////
    //////////////////////////////////
    
    public static CardDealer getInstance(){
        if(dealer == null){
            dealer = new CardDealer();
        }
        
        return dealer;
    }
    
    public Treasure nextTreasure(){
        if(unusedTreasures.isEmpty()){
            unusedTreasures = (ArrayList<Treasure>) usedTreasures.clone();
            usedTreasures.clear();
            shuffleTreasures();
        }
        Treasure aux = unusedTreasures.get(0);
        unusedTreasures.remove(0);
        
        return aux;
    }
    
    public Monster nextMonster(){
        if(unusedMonsters.isEmpty()){
            unusedMonsters = (ArrayList<Monster>) usedMonsters.clone();
            usedMonsters.clear();
            shuffleMonsters();
        }
        Monster aux = unusedMonsters.get(0);
        unusedMonsters.remove(0);
        
        return aux;
    }
    
    
    public Cultist nextCultist() {
        Cultist sectario;
        if(unusedCultists.isEmpty()){
            sectario = null; 
        }
        else {
            sectario = unusedCultists.get(0);
            unusedCultists.remove(0);
        }
        return sectario;
    }
    //--------------------------------------------------------------------
    
    public void giveTreasureBack(Treasure t){
        
        usedTreasures.add(t);
    }
    
    public void giveMonsterBack(Monster m){
        
        usedMonsters.add(m);
    }
    
    public void initCards(){
        initTreasureCardDeck();
        initMonsterCardDeck();
        initCultistCardDeck();
        
    }
    
    
}
